### PR TITLE
Add transpose kwargs for dot

### PR DESCRIPTION
**Pull request by jermainewang**
Also added the unittest.

### CODE REVIEW

File: src/operator/tensor/matrix_op-inl.h [hunk 1]
```diff
@@ -376,8 +399,19 @@ void DotBackward_(const nnvm::NodeAttrs& attrs,
     mshadow::Tensor<xpu, 2, real_t> mrhs_data = inputs[2].get<xpu, 2, real_t>(s);
     mshadow::Tensor<xpu, 2, real_t> mlhs_grad = outputs[0].get<xpu, 2, real_t>(s);
     mshadow::Tensor<xpu, 2, real_t> mrhs_grad = outputs[1].get<xpu, 2, real_t>(s);
-    ASSIGN_DISPATCH(mrhs_grad, req[1], dot(mlhs_data.T(), mout_grad));
-    ASSIGN_DISPATCH(mlhs_grad, req[0], dot(mout_grad, mrhs_data.T()));
+    if (param.transpose_a && param.transpose_b) {
```
Comment (piiswrong) [inline src/operator/tensor/matrix_op-inl.h]: are you sure these are correct?

Comment (jermainewang) [inline src/operator/tensor/matrix_op-inl.h]: Hmm... sorry I think I made some mistakes. We know that:
```
z = dot(x, y)
dx = dot(dz, y.T)
dy = dot(x.T, dz)
```
Then, we should have:
```
z1 = dot(x.T, y)
dx = dot(dz1, y).T
dy = dot(x, dz1)

z2 = dot(x, y.T)
dx = dot(dz2, y)
dy = dot(x.T, dz2).T

z3 = dot(x.T, y.T)
dx = dot(dz3, y).T
dy = dot(x, dz3).T
```
Basically, I miss the T here. Does  this make sense to you?

Question: how to do `dot(x, y).T` in one mshadow call? Is it equal to one blas call or one blas call + one physical transpose?

Comment (jermainewang) [inline src/operator/tensor/matrix_op-inl.h]: BTW, do we have any gradient checker test?

**Comment by piiswrong** [comment]
You cannot do dot xy T you need to do dot yT xT

We have gradient check see testoperator

Minjie Wang <notifications@github.com>于2017年1月2日 周一下午6:31写道：

**Comment by jermainewang** [comment]
Done the change.

**Pull request merged**